RecoEgamma/EgammaTools/plugins/EGRegressionModifierV3.cc

```diff
@@ -0,0 +1,364 @@
+  evt.getByToken(rhoToken_, rhoHandle);
+  rhoValue_ = *rhoHandle;
+}
+
+void EGRegressionModifierV3::setEventContent(const edm::EventSetup& iSetup)
+{
+  if(eleRegs_) eleRegs_->setEventContent(iSetup);
+  if(phoRegs_) phoRegs_->setEventContent(iSetup);
+  if(useClosestToCentreSeedCrysDef_) iSetup.get<CaloGeometryRecord>().get(caloGeomHandle_);
+}
+
+void EGRegressionModifierV3::modifyObject(reco::GsfElectron& ele) const
+{
+  //check if we have specified an electron regression correction and
+  //return the object unmodified if so
+  if(!eleRegs_) return;
+
+  const reco::SuperClusterRef& superClus = ele.superCluster();
+  
+  // skip HGCAL for now
+  if( superClus->seed()->seed().det() == DetId::Forward ) return;
+
+  const int numberOfClusters =  superClus->clusters().size();
+  const bool missingClusters = !superClus->clusters()[numberOfClusters-1].isAvailable();
+  if( missingClusters ) return ; // do not apply corrections in case of missing info (slimmed MiniAOD electrons)
+  //check if fbrem is filled as its needed for E/p combination so abort if its set to the default value 
+  //this will be the case for <5 (or current cuts) for miniAOD electrons
+  if(ele.fbrem()==reco::GsfElectron::ClassificationVariables().trackFbrem) return;
+  
+  auto regData = getRegData(ele);
+  const float rawEnergy = superClus->rawEnergy(); 
+  const float rawESEnergy = superClus->preshowerEnergy();
+  //bug here, it should include the ES, kept for backwards compat
+  const float rawEt = rawEnergy*superClus->position().rho()/superClus->position().r();
+  const bool isSaturated = ele.nSaturatedXtals()!=0;
+
+  const float ecalMean = eleRegs_->ecalOnlyMean(rawEt,ele.isEB(),isSaturated,regData.data());
+  const float ecalMeanCorr = ecalMean > 0. ? ecalMean : 1.0;
+  const float ecalSigma = eleRegs_->ecalOnlySigma(rawEt,ele.isEB(),isSaturated,regData.data());
+
+  const float corrEnergy = (rawEnergy + rawESEnergy)*ecalMeanCorr;
+  const float corrEnergyErr = corrEnergy*ecalSigma;
+  
+  ele.setCorrectedEcalEnergy(corrEnergy);
+  ele.setCorrectedEcalEnergyError(corrEnergyErr);
+
+  std::pair<float,float> combEnergyAndErr =  eleRegs_->epComb.combine(ele);
+  const math::XYZTLorentzVector newP4 = ele.p4()*combEnergyAndErr.first/ele.p4().t();
+  ele.correctMomentum(newP4, ele.trackMomentumError(), combEnergyAndErr.second);
+}
+
+void EGRegressionModifierV3::modifyObject(pat::Electron& ele) const
+{
+  modifyObject(static_cast<reco::GsfElectron&>(ele));
+}
+
+void EGRegressionModifierV3::modifyObject(reco::Photon& pho) const
+{
+  //check if we have specified an photon regression correction and
+  //return the object unmodified if so
+  if(!phoRegs_) return;
+  
+  const reco::SuperClusterRef& superClus = pho.superCluster();
+
+  // skip HGCAL for now
+  if(superClus->seed()->seed().det() == DetId::Forward ) return;
+
+  const int numberOfClusters =  superClus->clusters().size();
+  const bool missingClusters = !superClus->clusters()[numberOfClusters-1].isAvailable();
```

Comment:
likewise "back" method does not exist for this class

Reply:
this can also be a ```superClus->clusters().isAvailable()```
https://github.com/cms-sw/cmssw/blob/82e0ab0d145855d81d4da1251198d37dec744407/DataFormats/Common/src/PtrVectorBase.cc#L83-L99